src/Module.ts

```diff
@@ -369,20 +377,13 @@ export default class Module {
 	}
 
 	private analyse () {
-		enhance(this.ast, this, this.comments, this.dynamicImports);
-
-		// discover this module's imports and exports
-		let lastNode: Node;
-
+		enhance(this.ast, this, this.dynamicImports);
```

Comment:
If you're no longer using `this.comments` does that mean this property can be removed entirely, including the use of the `onComment` hook?

Reply:
Just tested this and the answer is yes :)

Reply:
Unfortunately not as it is still required to remove source map comments. This, however, is the only reason why we keep this.

Reply:
Ahh. Can you point me to the code?

Reply:
Maybe at some point we should implement our own algorithm to scan for and remove those comments.

Reply:
I still can't locate any other references to `module.comments` anywhere in the codebase, so was just wondering where this code lives?

Reply:
Also this PR then solves a lot of the problems of comments for provided ASTs!

Reply:
It's here: https://github.com/rollup/rollup/blob/82fe34dfe39d6c2aea9ca32444ea4a68e3afe814/src/Module.ts#L232

Reply:
It's a side-effect of the filtering

Reply:
I guess we could just scan for them via `code.original.indexOf('//#')` and then remove until the next `\n`. `indexOf` seems to be really fast.

Reply:
Ah I see. A sourceMappingURL comment can only be a block comment so it's fine to do something custom. Also it can be based on `lastIndexOf` as a backwards search through the string would be quicker.

Here's the code I have used for this previously - https://github.com/systemjs/builder/blob/master/lib/sourcemaps.js#L12.

Reply:
Looking at this example, maybe the issue is not that trivial and we should probably rather trust acorns parsing here and leave it as it is: https://github.com/rollup/rollup/blob/22aa1fb7541deb3d3a6b305898902ccb06b7d0c2/test/form/samples/removes-existing-sourcemap-comments/main.js#L3

Reply:
Otherwise we might break the sourceMap handling of other libraries.

Reply:
Here's v8's method to do this - https://github.com/v8/v8/blob/8d38c15e04598f9e48e6230327da0e41a2445957/src/inspector/search-util.cc#L16

Note that it doesn't use a lexer at all.

I think it would be much simpler to make this code its own method and deprecate `this.comments` entirely. But happy to make this a separate PR too if you prefer.

Reply:
Thinking about this further detection is a less stringent problem than removal - and that’s what I got confused here. It makes sense that Rollup has to hold to a higher standard for removal not to break code.

Perhaps we can inline the removal into the parsing code without separately gathering the comments at least.